### ISSUE TITLE
Universal default shader? 

### ISSUE THREAD

**Issue by dragonCASTjosh** [urho3d/Urho3D#1566]
Is there a benefit to having the maze of different techniques compared to having a simple shader that can be dynamically changed from the material editor, for example if you add a normal map it will add the define for you in the material. The reason this comes to mind is i have been playing around with other engines and they all tend to stick to a standard shader system rather then techniques. 

In short we use the 1 shader per renderpath we have now but change the defines from the material, kinda like Unitys standard shader from example.

This would also allow a lot easier options for the user for example possibly a tick box for PBR, and a drop down for the diffuse model. If given help on the editor side then im more then happy to work on a standard shader.

**Comment by cadaver** [comment]
The intention of the Technique is to hide technical details of the passes, e.g. base pass has its own blend setting, additive light pass has its own. I don't think this can be removed entirely or else it would have to be written somewhere else. Or alternatively materials would get a lot more complex.

However what could make sense would be to allow the material to set additional shader defines, in which case the technique would be dynamically cloned, and in that case we wouldn't need as many techniques.

**Comment by dragonCASTjosh** [comment]
Would it be possible to replace the Technique stuff with passes in the material editor. So for example where currently you can add multiple Techniques it would be come the passes you wish to perform and the order you want then done?

**Comment by dragonCASTjosh** [comment]
I think it would also help for if someone wants to add a new pass to the list for example Sub-Surface they could just write the shader and add a pass xml and it would be in the list.

**Comment by cadaver** [comment]
The pass order is determined by the renderpath actually, so you can't change that from the material.

If you wanted to get rid of techniques, you would need to solve where the information like "base pass with replace blend mode, additional light pass with additive blend mode" would be stored. If this has to be set & repeated for every material using a certain shader it's slightly hardcore. On the other hand, the same shader can be used with multiple renderstate settings.

If you want to change this, feel free to work on a PR, but we shouldn't lose functionality.

**Comment by dragonCASTjosh** [comment]
I will look into it see what i can come up with

**Comment by 1vanK** [comment]
Technics are good for those who do not know the engine (just use technic in material and job is done). But to create your materials uncomfortable with your shaders.

How about this compromise:
All passes are placed in the material file, but it is possible to incude another file into material (I am not sure that is possible for xml)
```
<material>
   #include MaterialTemplates/Diff.template
   <parameter name="MatSpecColor" value="0.5 0.5 0.5 16" />
</material>
```
Also, I do not know how to make LODs for materials

**Comment by dragonCASTjosh** [comment]
I like the idea of templates for the core techniques. but maybe have it as Diff handles all the standard shading and from that the material editor will add defines for normal maps and such only if used.

**Comment by 1vanK** [comment]
Editor can not include nothing, just place all material passes to file

**Comment by dragonCASTjosh** [comment]
not sure what you mean by that

**Comment by cadaver** [comment]
In the interest of generality I don't think the material editor should have hardcoded intelligence for texture units and shader defines.

Unity actually has quite nice shader concept where the normal mapped shader code is different than plain diffuse. However, that could lead to combination explosion as well, considering e.g. PBR and non-PBR shaders.

**Comment by 1vanK** [comment]
Sorry, English hard for me :) I mean when user create new material in Editor, Editor not include templates to material file, just necessary passes

**Comment by dragonCASTjosh** [comment]
I dont think the uniforms and textures units are going to change it would be a case of having the editor write a define in the material for the texture. 

Maybe the template system could start with the minimum required passes for a shader to work and they if you want to change it you can select from a list of template for example "Standard Alpha" or "Standard PBR"

**Comment by cadaver** [comment]
One thing you also need to consider is how deferred / forward renderpath switches are going to be handled. Currently Technique defines everything needed for both forward, prepass and deferred, but to have these in every material will lead to quite cluttered looking materials.

It's actually easier in a project where you decide to use only deferred, or only forward, and are writing your own techniques, which will then end up simpler. But Urho itself (at least the materials/techniques that come with the engine) cannot make that choice.

**Comment by dragonCASTjosh** [comment]
Essentially it would be automatic creation of techniques based on what the user is using. I think that if we treat the techniques we have now as a generalized system for example the "Standard" technique will handle all materials that are just opaque dynamic defines for Diffuse, Normal, Secular and Emmisive textures.

**Comment by eugeneko** [comment]
I have the same problem.. If I want to create my own ubershader that handles everything, I have to create tons of techniques with upper limit

`[NumberOfGeometryTypes]x3[Opaque/AlphaMask/AlphaBlend]x6[NoTexture/Diff/DiffNormal/DiffNormalMetallic/DiffNormalRoughness/DiffNormalMetallicRoughness]x???[Who knows what else?]>=18N`

**Comment by cadaver** [comment]
You don't need techniques for geometry types, since that's automatically handled by the engine. But shader parameters from material looks like a good first step for reducing the problem.

**Comment by eugeneko** [comment]
@cadaver 
I mean *really* different types of geometry those *need* different shaders and techniques.
Maybe we need some new entity like TechniqueGroup that could describe set of needed techniques in data-driven way? When applied to material, it is 'converted' to actual techniques depending on resources that are actually set and some material flags like alpha blending info.
Can material be allowed to expose vs/psdefiness too? It should also reduce this problem.

**Comment by cadaver** [comment]
Yes that's what I was suggesting. When a material with its own defines gets loaded, it would clone the techniques, then inject its own defines.

**Comment by eugeneko** [comment]
@cadaver 
Huh, I am still not very familiar with Urho internals... If Technique is just cloned inside material with some new define (e.g ALPHAMASK), it couldn't be simply re-used by some another material because it has no resource in cache. Actually, this cloned Technique could be put into ResourceCache with some synthetic name like PBRDiff.xml/ALPHAMASK, but it looks a bit dirty...
Finally, is it so important to reuse Technique-s instances?

**Comment by cadaver** [comment]
It would be preferable. Alternative is to make each Technique manage its own clones with added defines.

**Comment by eugeneko** [comment]
Huh. Seems that this simple improvement already covers a lot of cases.. I've tested it on simple ALPHAMASK overriding.
https://github.com/eugeneko/Urho3D/commit/e3538f3892a9b2f3bfb7f7181496f7df66bbc442

**Comment by cadaver** [comment]
It could, but I don't think it's super critical, as it's often used only at material load time.

Btw. I'm making my own implementation which I'll commit in very near future, so I don't recommend working on your own at the moment. I'm at the same time removing features which no longer can be accurately tracked, namely the alpha masking hint and automatic determination of whether material should occlude. That can be manually set instead.

**Comment by eugeneko** [comment]
I've just wanted to test things are mentioned here.
Can you briefly explain what are you doing, please? I mean in the context of this issue. Do you have propper solution against exploding techniques?

**Comment by cadaver** [comment]
Added API to Material to set shader defines. This should be enough for the case where the shader defines threaten to make technique count explode. Other things like blend modes, and defining different shaders for different passes still have to go through the actual Technique.

**Comment by cadaver** [comment]
Closing this since no further discussion for some time. New issue or PR can be opened if there are concrete ideas for changes.
